Make StripeSCA payment method work with existing credit cards (the ones saved through the Stripe Charges API)

luisramos0 opened the pull request:
#### What? Why?

Closes #4685
Here we make all cards stored with StripeSCA go to the platform account (which is what happens with the existing Stripe payment method). To use these cards in the Payment Intents API we need to clone these cards to the hubs stripe accounts (the connected accounts) before we use them. This process of cloning will also work for existing cards saved through the Charges API :+1:
So, this PR makes this payment method complete and compatible with the existing cards in the system.

#### What should we test?
First verify that #4672 is working by following it's test guide.
Afterwards, see acceptance criteria for this PR in #4685.

#### Release notes
Changelog Category: Changed
Made new Stripe SCA payment method compatible with credit cards already stored in Stripe. This payment method is now complete and we can now start to migrate all users to this payment method.

mkllnk commented on lib/stripe/card_cloner.rb:
```diff
@@ -0,0 +1,44 @@
+# This process is used in the migration between using the Stripe Charges API (stripe_connect)
+#   and the Stripe Payment Intents API (stripe_sca)
+#
+# This process can be deleted once all hubs are running on the new stripe_sca method and all cards in the system have been migrated to the new payment_methods
```
I'm reviewing this on Github and need to scroll horizontally which is a bit awkward. Can I have shorter lines, please?

luisramos0 replied:
yes, this was fixed in later commits :+1:

mkllnk commented on lib/stripe/card_cloner.rb:
```diff
@@ -0,0 +1,44 @@
+
+      credit_card.update_attributes gateway_customer_profile_id: new_customer.id,
+                                    gateway_payment_profile_id: new_payment_method.id
+      credit_card
```
You don't need a return value here, do you?

Once we converted a card to the new API, we can't switch back to Stripe Connect, can we?

luisramos0 replied:
this was changed in later commits.

This process clones a card in a platform account to the connected account.
This process is backward compatible! Because the original stripe method will take the card in a platform account and create a "token" in the connected account, this method takes the card in the platform account and create another card in the connected account.

mkllnk commented on spec/lib/stripe/credit_card_cloner_spec.rb:
```diff
@@ -0,0 +1,42 @@
+      before do
+        allow(Stripe).to receive(:api_key) { "sk_test_12345" }
+
+        stub_request(:post, "https://api.stripe.com/v1/customers")
```
I'm wondering if it's time for the vcr gem. We used it for Fair Food and it's working really well. It records real API responses to repeat them offline. So you get real data and you are setup to repeat it against the live API whenever you want.

luisramos0 replied:
I like this webmock way of testing things but @kristinalim will be looking into testing these things, maybe a good pointer for #4639

Matt-Yorkley commented on lib/stripe/credit_card_cloner.rb:
```diff
@@ -0,0 +1,52 @@
+#
+# We are doing this process every time the card is charged:
+#   - this means that, if the customer uses the same card on the same seller multiple times,
+#       the card will be created multiple times on the seller's account
```
I've got a couple of questions, just out of interest:

- What are the consequences of the same card being created multiple times on the seller's account?
- If a card already exists on the seller's account, can't we retrieve it using the Stripe API instead of recreating it?

luisramos0 replied:
1. until now what was created every time on the seller's account was the token, there were as many tokens as payments on these accounts. the stripe dashboard doesnt show the tokens I think. Now we are doing the same for cards, this appears in the dashboard on the customers tab, each time there is a payment there will be a new customer, even if the card has been previously used. I dont think this is a big problem, it's just a detail in the stripe UI for the sellers.
2. yes we can, it's just more code and more work to get that done. I dont think there's much value in implementing that.

filipefurtad0 commented:
Hey @luisramos0,

Performing the first step of the testing notes:
1. make a payment without saving card

Generated the error:

_No such PaymentMethod: pm_1GEbg1LE3KX9OdcFl4vKeSul; OAuth key or Stripe-Account header was used but API request was provided with a platform-owned payment method ID. Please ensure that the provided payment method matches the specified account._ 

Please see testing notes:
https://docs.google.com/document/d/1wQzeV_NB-QaFF5dCrG4bBx-7lve66vc0o1ZRP-SWjQY/edit#

I'll stop testing now, so removing the staging-label.

luisramos0 commented:
🙈 I messed up my local branches between my local linux and my local max.... I was lucky and managed to recover the lost code. It was only one commit to fix exactly the use case you tested:
https://github.com/openfoodfoundation/openfoodnetwork/pull/4719/commits/404e7c1f37127070f5cbf151cc6694c54e316ae1

let's just watch the new build and then stage and test again 👍

luisramos0 commented:
ok, looking good, I restaged in katuma and managed to checkout with stripe on fabulous funcho :-)

filipefurtad0 commented:
Hi @luisramos0,

Awsome! This looks very good indeed. :tada: 

Testing notes:
https://docs.google.com/document/d/1wQzeV_NB-QaFF5dCrG4bBx-7lve66vc0o1ZRP-SWjQY/edit#

I do think there is a small issue with the checking of payment details, of a given order, for the new Stripe SCA. But since this is not the focus of the PR, and all the test cases worked fine, I am moving this to ready to go.

Thanks!

lin-d-hop commented:
Hiya @filipefurtad0 

Thanks so much for testing this!

I'm a little nervous about introducing regressions. Did you check in with others that these aren't related to this PR? Seems if inspecting individual payments work on prod and not here we might have a problem?

A user facing slug is pretty bad. Would be great to avoid if possible. Though I'm not sure if the pages with slugs are used much (I doubt it).

My personal preference would be not to introduce these slugs. Though if @luisramos0 and/or @RachL are happy then I'm happy.

RachL commented:
I've hit the same issue in a previous PR concerning some spree backend migration, maybe it is linked?

Worth to have it checked I believe.

filipefurtad0 commented:
@lin-d-hop and @RachL,
Thank you for your feedback on this!
Definitively worth a second look - to avoid having these slugs later on and at least to make sure these are not related to this PR, in particular. Will get back at this.
Thanks again :-)

luisramos0 commented:
great testing @filipefurtad0 
snail 1 is a bug in this PR - no template to show stripeSCA payment details - this template was now added
snail 2 is a pre-existing bug - payment details template could not handle nil credit card (deleted by user for example), it is now able to handle that case, it will show empty credit card details and the payment amount.
Added 2 commits to fix these two problems and one auto spec for each case :muscle: 

Changes are minor, @Matt-Yorkley or @mkllnk could you have a quick look please at these two last commits?
Thanks!

luisramos0 commented:
Thanks @mkllnk 
Moving to Test Ready again.

filipefurtad0 commented:
Hi @luisramos0,

I re-tested all test cases:
- the test-guide from #4672
- and acceptance criteria for this PR, as in #4685

Also checked for both Snail-cases on the test notes.
All good, this is ready-to-go!

luisramos0 commented:
great, thanks Filipe!!!
